modules/extension/app-schema/app-schema/src/main/java/org/geotools/data/complex/IndexedMappingFeatureIterator.java

```diff
@@ -0,0 +1,228 @@
+
+    protected final AppSchemaDataAccess store;
+    protected final FeatureTypeMapping mapping;
+    protected final Query query;
+    protected final Filter unrolledFilter;
+    protected final Transaction transaction;
+
+    protected IndexIdIterator indexIterator;
+    protected FeatureIterator<? extends Feature> sourceIterator;
+
+    protected IndexQueryManager indexModeProc;
+    protected QueryIndexCoverage queryMode;
+
+    public IndexedMappingFeatureIterator(
+            AppSchemaDataAccess store,
+            FeatureTypeMapping mapping,
+            Query query,
+            Filter unrolledFilter,
+            Transaction transaction,
+            IndexQueryManager indexModeProcessor) {
+        this.store = store;
+        this.mapping = mapping;
+        this.query = query;
+        this.unrolledFilter = unrolledFilter;
+        this.transaction = transaction;
+        this.indexModeProc = indexModeProcessor;
+        selectExecutionPlan();
+    }
+
+    /**
+     * Analyze query and select a plan: 1.- All fields indexed, execute all query on index layer 2.-
+     * Mixed fields indexed and not, execute indexed operators and re-map query to database
+     */
+    protected void selectExecutionPlan() {
+        queryMode = indexModeProc.getIndexMode();
+    }
+
+    protected Query unrollIndexes(Query query) {
+        Query newQuery = new Query(query);
+        newQuery.setFilter(unrollFilter(query.getFilter()));
+        return newQuery;
+    }
+
+    protected SortBy[] unrollSortBy(SortBy[] sortArray) {
+        if (sortArray == null) return null;
+        ArrayList<SortBy> unrolledSorts = new ArrayList<>();
+        FilterFactory2 ff = CommonFactoryFinder.getFilterFactory2(GeoTools.getDefaultHints());
+        for (SortBy aSort : sortArray) {
+            SortBy newSort =
+                    ff.sort(
+                            unrollIndex(aSort.getPropertyName(), mapping).getPropertyName(),
+                            aSort.getSortOrder());
+            unrolledSorts.add(newSort);
+        }
+        return unrolledSorts.toArray(new SortBy[] {});
+    }
+
+    public static PropertyName unrollIndex(PropertyName expression, FeatureTypeMapping mapping) {
+        String targetXPath = expression.getPropertyName();
+        AttributeMapping attMp = getIndexedAttribute(mapping, targetXPath);
+        if (attMp != null) {
+            return new AttributeExpressionImpl(attMp.getIndexField());
+        }
+        return new AttributeExpressionImpl((String) null);
+    }
+
+    protected Filter unrollFilter(Filter filter) {
+        IndexUnmappingVisitor visitor = new IndexUnmappingVisitor(mapping);
+        Filter unrolledFilter = (Filter) filter.accept(visitor, null);
+        return unrolledFilter;
+    }
+
+    /**
+     * Simplifies id value, cutting "typename." part if exists
+     *
+     * @param feature
+     * @return
+     */
+    protected String simplifyIndentifier(Feature feature) {
```

Comment:
I see this one also in IndexIdIterator, maybe we can factor it out for reuse

Reply:
Ah! Nice catch ! This method was _already_ factored out indeed, I just frogot to remove it from here :yum:.